app/src/main/java/fr/free/nrw/commons/upload/worker/UploadWorker.kt

```diff
@@ -402,10 +422,33 @@ class UploadWorker(var appContext: Context, workerParams: WorkerParameters) :
                 }
                 else -> {
                     Timber.e("""upload file to stash failed with status: ${stashUploadResult.state}""")
-                    showFailedNotification(contribution)
+                    showInvalidLoginNotification(contribution)
                     contribution.state = Contribution.STATE_FAILED
                     contribution.chunkInfo = null
                     contributionDao.saveSynchronous(contribution)
+                    if (stashUploadResult.errorMessage.equals(CsrfTokenClient.INVALID_TOKEN_ERROR_MESSAGE)) {
+                        Timber.e("Invalid Login, logging out")
+                        val username = sessionManager.userName
+                        sessionManager.logout()
+                            .andThen(Completable.fromAction {
+                                Timber.d("All accounts have been removed")
+                                clearImageCache()
+                                updateAllDatabases()
+                            })
+                            .subscribeOn(Schedulers.io())
+                            .observeOn(AndroidSchedulers.mainThread())
+                            .subscribe(
+                                {
+                                    val intent = Intent(appContext, LoginActivity::class.java)
+                                    intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
+                                    intent.putExtra(loginMessage, appContext.getString(R.string.invalid_login_message))
+                                    intent.putExtra(loginUsername,username)
+                                    appContext.startActivity(intent)
+                                }
+                            ) { t: Throwable? -> Timber.e(t) }
+                    } else {
```

Comment:
This whole piece and the related methods seem like a duplicate of what we already have in the CommonsApplication.clearApplicationData. Can we avoid duplication in the code by extracting the redundant parts? I suppose that this would be especially helpful considering we're planning on doing this handling in other places too.

Here's a quick-and-dirty commit to show what I have in mind: https://github.com/sivaraam/apps-android-commons/commit/a2d5a9d0df5a0e9f8f9d5eadd5cb46ea777c7fdb

Let me know in case I'm missing something.